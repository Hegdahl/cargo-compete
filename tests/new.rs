pub mod common;

use ignore::overrides::Override;
use insta::{assert_json_snapshot, assert_snapshot};
use liquid::object;
use std::io::BufRead;

#[test]
fn atcoder_agc047() -> anyhow::Result<()> {
    let (output, tree) = run(&b""[..], "agc047")?;
    assert_snapshot!("atcoder_agc047_output", output);
    assert_json_snapshot!("atcoder_agc047_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

#[cfg(feature = "__test_with_credentials")]
#[test]
fn atcoder_practice() -> anyhow::Result<()> {
    let (output, tree) = run(common::atcoder_credentials()?, "practice")?;
    assert_snapshot!("atcoder_practice_output", output);
    assert_json_snapshot!("atcoder_practice_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

fn run(
    input: impl BufRead + 'static,
    contest: &str,
) -> anyhow::Result<(String, serde_json::Value)> {
    common::run(
        |workspace_root| -> _ {
            std::fs::create_dir_all(workspace_root.join("cargo-compete-template").join("src"))?;

            std::fs::write(
                workspace_root.join("Cargo.toml"),
                r#"[workspace]
members = ["cargo-compete-template"]
"#,
            )?;

            std::fs::write(
                workspace_root.join("compete.toml"),
                liquid::ParserBuilder::with_stdlib()
                    .build()?
                    .parse(include_str!("../resources/compete.toml.liquid"))?
                    .render(&object!({
                        "template_platform": "atcoder",
                        "submit_via_binary": false,
                    }))?,
            )?;

            std::fs::write(
                workspace_root
                    .join("cargo-compete-template")
                    .join("Cargo.toml"),
                r#"[package]
name = "cargo-compete-template"
version = "0.1.0"
edition = "2018"
"#,
            )?;

            std::fs::write(
                workspace_root
                    .join("cargo-compete-template")
                    .join("src")
                    .join("main.rs"),
                r#"fn main() {
    todo!();
}
"#,
            )?;
            Ok(())
        },
        input,
        &["", "compete", "n", contest],
        |workspace_root, output| {
            output
                .replace(workspace_root.to_str().unwrap(), "{{ cwd }}")
                .replace('/', "{{ slash_or_backslash }}")
                .replace('\\', "{{ slash_or_backslash }}")
        },
        |_| Ok(Override::empty()),
    )
}
