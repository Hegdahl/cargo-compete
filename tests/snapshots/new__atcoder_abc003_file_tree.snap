---
source: tests/new.rs
expression: tree
---
{
  ".cargo": {
    "config.toml": "[cargo-new]\nname = \"\"\nemail = \"\"\n"
  },
  "abc003": {
    "Cargo.toml": "[package]\nname = \"abc003\"\nversion = \"0.1.0\"\nauthors = [\"\"]\nedition = \"2018\"\n\n[package.metadata.cargo-compete.bin]\na = { name = \"abc003-a\", problem = \"https://atcoder.jp/contests/abc003/tasks/abc003_1\" }\nb = { name = \"abc003-b\", problem = \"https://atcoder.jp/contests/abc003/tasks/abc003_2\" }\nc = { name = \"abc003-c\", problem = \"https://atcoder.jp/contests/abc003/tasks/abc003_3\" }\nd = { name = \"abc003-d\", problem = \"https://atcoder.jp/contests/abc003/tasks/abc003_4\" }\n\n[[bin]]\nname = \"abc003-a\"\npath = \"src/bin/a.rs\"\n\n[[bin]]\nname = \"abc003-b\"\npath = \"src/bin/b.rs\"\n\n[[bin]]\nname = \"abc003-c\"\npath = \"src/bin/c.rs\"\n\n[[bin]]\nname = \"abc003-d\"\npath = \"src/bin/d.rs\"\n[dependencies]\nproconio = \"=0.3.6\"\n",
    "src": {
      "bin": {
        "a.rs": "fn main() {\n    todo!();\n}\n",
        "b.rs": "fn main() {\n    todo!();\n}\n",
        "c.rs": "fn main() {\n    todo!();\n}\n",
        "d.rs": "fn main() {\n    todo!();\n}\n"
      }
    },
    "testcases": {
      "a.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch:\n  Float:\n    relative_error: 0.000001\n    absolute_error: 0.000001\n\ncases:\n  - name: sample1\n    in: |\n      6\n    out: |\n      35000\n  - name: sample2\n    in: |\n      91\n    out: |\n      460000\n\nextend: []\n",
      "b.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      ch@ku@ai\n      choku@@i\n    out: |\n      You can win\n  - name: sample2\n    in: |\n      aoki\n      @ok@\n    out: |\n      You will lose\n  - name: sample3\n    in: |\n      arc\n      abc\n    out: |\n      You will lose\n\nextend: []\n",
      "c.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch:\n  Float:\n    relative_error: 0.000001\n    absolute_error: 0.000001\n\ncases:\n  - name: sample1\n    in: |\n      2 2\n      1000 1500\n    out: |\n      1000.000000\n  - name: sample2\n    in: |\n      2 1\n      1000 1500\n    out: |\n      750\n  - name: sample3\n    in: |\n      10 5\n      2604 2281 3204 2264 2200 2650 2229 2461 2439 2211\n    out: |\n      2820.031250000\n\nextend: []\n",
      "d.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      3 2\n      2 2\n      2 2\n    out: |\n      12\n  - name: sample2\n    in: |\n      4 5\n      3 1\n      3 0\n    out: |\n      10\n  - name: sample3\n    in: |\n      23 18\n      15 13\n      100 95\n    out: |\n      364527243\n  - name: sample4\n    in: |\n      30 30\n      24 22\n      145 132\n    out: |\n      976668549\n\nextend: []\n"
    }
  },
  "compete.toml": "test-suite = \"{{ manifest_dir }}/testcases/{{ bin_alias | kebabcase }}.yml\"\n\n[new]\nplatform = \"atcoder\"\npath = \"./{{ package_name }}\"\n\n[new.template.dependencies]\nkind = \"inline\"\ncontent = '''\nproconio = \"=0.3.6\"\n'''\n\n[new.template.src]\nkind = \"inline\"\ncontent = '''\nfn main() {\n    todo!();\n}\n'''\n"
}
